Store taffy node keys in components

ickshonpe opened the pull request:
# Objective

Store the taffy node keys in a component on the UI node entities instead of using the `UiSurface::entity_to_taffy` map to look them up.

## Changelog
`UiKey`:
* `UiKey` is a new component with a single field: `taffy_id: taffy::node::Node` and a method `get` to retrieve its value.
* `UiKey::default()` creates a `UiKey` with a null taffy id value. This should be the only way for users to instantiate a `UiNodeId` component. If multiple entities with the same non-null `UiKey::taffy_id` exist, the UI layout will not update correctly.
* All UI nodes must have a `UiKey` component to be included in the UI layout.

`NodeBundle`, `ImageBundle`, `TextBundle`, `ButtonBundle`:
* The `UiKey` component has been added to each of these bundles.

`UiSurface`:
* The `upsert_node` method has been split into separate methods: `insert_node` and `set_style`. 
* Removed `get_layout`. The Taffy layout is accessed directly instead.
* Most methods now accept a `taffy::node::Node` parameter instead since `entity_to_taffy` lookups are no longer needed.

`LayoutError`
* Removed as no longer used, neither error should ever occur.

`ui_layout_system`
* Removal detection is performed first before updates. Otherwise, the UI node entity tree and the Taffy layout tree could get out of sync if a user removes and then, later in the same frame, reinserts a component used by the UI.

## Migration Guide
`UiKey` is a new component that is managed automatically by `ui_layout_system` and should only be instantiated with `UiKey::default()`. It is used to identify a UI node in the UI's internal layout tree. 

`UiKey` has been added to each of the UI node bundles.

ickshonpe committed:
changes:

* Added new components "TaffyNode" and "TaffyParent"
* Added `insert_node` method to `FlexSurface`
* Added `make_measure` function to `flex` module.
* `flex_node_system` stores taffy nodes keys in the "TaffyNode" and "TaffyParent" components instead of looking them up using the `entity_to_taffy` map.

ickshonpe commented:
```
cargo run --profile stress-test --example many_buttons --features trace_tracy -- no-text
```

<img width="295" alt="flex_node_system_tracy" src="https://user-images.githubusercontent.com/27962798/229600478-520d7701-60e6-4df0-9327-a80095d0f82d.PNG">

james7132 commented on crates/bevy_ui/src/flex/mod.rs:
```diff
@@ -22,6 +24,16 @@ use taffy::{
 };
 
+#[derive(Component, Copy, Clone)]
+pub struct TaffyNode {
```
Could we collapse these two components into `TaffyState` or just merge it into `Node`? Unless there's a good reason (i.e. parallelism, ambiguity, cache bandwidth) to keep these as separate components, I'm not sure it's worth it to introduce two more components.

ickshonpe replied:
They can't go in `Node` I think because it's queried for size change detection.
Maybe they can be combined into a single component though. They aren't components that users should ever need to be aware of and they aren't part of any bundles.

ickshonpe replied:
Just realised, I think `TaffyParent` is only required because we're still using 'Transform` and it can be removed when we switch to a UI-specific transform.

nicoburns replied:
+1 for combining into a single component. I see no reason for these to be separate and combining them will save a lookup in the case you need both.

ickshonpe committed:
Remove `TaffyNode` and `TaffyParent` components when `Node` is removed.

ickshonpe committed:
changes:

* added `primary_node` and `physical_to_logical_factor` fields to `FlexSurface`.
* added the `Transforms` variant to the `UiSystem` enum.
* split off the transforms update from `flex_node_system` into a separate function `update_ui_node_transform`
and scheduled it to run after `UiSystem::Flex` and before `TransformSystem::TransformPropagate`.

ickshonpe committed:
changes:
* removed `TaffyParent` component, added `parent` field to `TaffyNode` component
* removed `PrimaryNode` field from `FlexSurface`

mockersf commented:
Is it possible to remove the `entity_to_taffy` hashmap? Otherwise it would mean storing the same information twice

ickshonpe commented:
> Is it possible to remove the `entity_to_taffy` hashmap? Otherwise it would mean storing the same information twice

I don't think so, it's needed for handling children and removing orphaned taffy nodes.

nicoburns commented:
> > Is it possible to remove the `entity_to_taffy` hashmap? Otherwise it would mean storing the same information twice
> 
> I don't think so, it's needed for handling children and removing orphaned taffy nodes. Maybe if Taffy could be made to store the layout data directly in the Bevy UI entities instead of using SlotMap.

https://github.com/DioxusLabs/taffy/pull/326 (or more likely, parts of this PR split out into their own PRs) will allow this. It's supposed to be possible with the current version of the [LayoutTree](https://docs.rs/taffy/latest/taffy/tree/trait.LayoutTree.html) trait, but in practice the current version has things like SlotMap keys in the trait function signatures which make it impractical.

With that approach *all* Taffy data (input styles, parent-children relationships, and and output layouts) would be stored by Bevy (however it chose to do so). Care would need to be taken to ensure performance parity though, as Taffy's SlotMaps are backed by Vecs and are therefore very fast to access. Not sure if it's possible to get Vec backed storage in bevy_ecs that only applies to UI nodes?

nicoburns commented:
Although to this specific issue there might be a simpler solution. My understanding is that a Taffy component typically *is* a HashMap, so a Query of a component containing the Taffy node key ought to function the same as a HashMap containing the mapping I think.

ickshonpe committed:
changes:
* Renamed `Node` to `NodeSize`
* Renamed `TaffyNode` to `Node` and removed its `Parent` field.
* Split off parts of `flex_node_system` into separate systems:
`synchronise_children`, `insert_new_ui_nodes_into_layout` and `clean_up_removed_ui_nodes`.

ickshonpe committed:
Merge branch 'main' into store-taffy-keys-in-components

ickshonpe committed:
cargo fmt

ickshonpe committed:
handle window events for ui in a separate system

ickshonpe committed:
fix scale factor changed check

ickshonpe committed:
cargo fmt

ickshonpe committed:
move window update back into `flex_node_system`

ickshonpe committed:
* Renamed `FlexSuface` to `UiSurface`.
* Renamed `flex_node_system` to `update_ui_layout`.
* Replaced the `window_nodes` map field of  `UiSurface` with a single `window_node` node field.

ickshonpe committed:
fixed tests

ickshonpe committed:
Cargo fmt --all

ickshonpe committed:
doc comment and clippy fixes

ickshonpe committed:
changes:
* Removed redundant `Node` component removal.
* Query for `Node` components of root nodes instead of `Entity` so that `set_window_children` doesn't have to look up the keys.
* Added a a simple test for node insertion and removal.

ickshonpe committed:
added another test

ickshonpe committed:
replaced `scale_factor_changed` and `resized` fields with a single `require_full_update` field

ickshonpe commented:
> > > Is it possible to remove the `entity_to_taffy` hashmap? Otherwise it would mean storing the same information twice
> > 
> > 
> > I don't think so, it's needed for handling children and removing orphaned taffy nodes. Maybe if Taffy could be made to store the layout data directly in the Bevy UI entities instead of using SlotMap.
> 
> [DioxusLabs/taffy#326](https://github.com/DioxusLabs/taffy/pull/326) (or more likely, parts of this PR split out into their own PRs) will allow this. It's supposed to be possible with the current version of the [LayoutTree](https://docs.rs/taffy/latest/taffy/tree/trait.LayoutTree.html) trait, but in practice the current version has things like SlotMap keys in the trait function signatures which make it impractical.
> 
> With that approach _all_ Taffy data (input styles, parent-children relationships, and and output layouts) would be stored by Bevy (however it chose to do so). Care would need to be taken to ensure performance parity though, as Taffy's SlotMaps are backed by Vecs and are therefore very fast to access. Not sure if it's possible to get Vec backed storage in bevy_ecs that only applies to UI nodes?

I tried to get a version working with generic keys but got lots of ICE errors, so I hacked together a very niave implementation with World implementing LayoutTree etc and it was horribly slow and unusable.

ickshonpe commented:
Worked on this quite a bit more, made a lot of changes:

* Removed any need for caching parent keys.
* Renamed `Node` to `NodeSize`.
* `TaffyKey` is renamed to `Node` and should be the primary component that all UI systems query for. 
* `TaffyKey` (now `Node`) is included in the UI node bundles instead of being added by the UI systems. The key is initially set to null and then assigned a permanent taffy key in the `insert_new_ui_nodes` system.
* `flex_node_system` is split up into multiple systems. Might be too granular now and some of them could be consolidated.
* The UI schedule can be run without needing a window or rendering.
* Added a couple of simple tests adding and removing nodes and styles.

ickshonpe committed:
Renamed `insert_new_ui_nodes_into_layout` to `insert_new_uinodes` and
`synchonise_children` to `synchronise_ui_children`.

ickshonpe committed:
cargo fmt

ickshonpe committed:
Merge branch 'main' into store-taffy-keys-in-components

ickshonpe committed:
fix errors from merge

ickshonpe committed:
Fixed remaining errors from merge

ickshonpe committed:
fix imports

ickshonpe committed:
use a parallel iterator to update UI transforms

ickshonpe committed:
Merge branch 'main' into store-taffy-keys-in-components

ickshonpe committed:
fixed errors from merge

ickshonpe committed:
cargo fmt

ickshonpe committed:
fix doc comments

ickshonpe committed:
fixed needless borrows

ickshonpe committed:
more doc comment fixes

ickshonpe committed:
Renamed `remove_entities` to `remove_nodes`

ickshonpe committed:
Merge branch 'main' into store-taffy-keys-in-components

ickshonpe committed:
* Fixed bugs in `debug` and `accessibility` caused by merge.

ickshonpe committed:
cargo fmt --all

ickshonpe committed:
Merge branch 'main' into store-taffy-keys-in-components

ickshonpe committed:
changes:
* Renamed `Node` to `NodeSize`
* Created a new `Node` component that holds an index to the corresponding Taffy node in `UiSurface`.
* Added `Node` to the UI's node bundles.
* Added a new system `insert_new_ui_nodes_system` that runs in `PostUpdate` before `ui_layout_system` that inserts new UI nodes into the taffy layout tree. Removed the `get_layout`, `upsert_leaf` and `update_measure` methods from `UiSurface`.
* Change `ui_layout_system` to use `Node` to look up taffy nodes where possible.

ickshonpe committed:
changes:
Changed queries to use `With<Node>`.

ickshonpe committed:
changes:
* renamed `insert_new_ui_nodes_system` to `insert_ui_nodes_system`
* moved node removal to `insert_ui_nodes_system`
* renamed `taffy_key` field to `taffy_node`

ickshonpe committed:
changes:
* renamed `taffy_key` field to `taffy_node`
* renamed `insert_new_ui_nodes_system` to `insert_ui_nodes_system`
* moved node removal to `insert_ui_nodes_system`
* remove orphaned measure funcs before measure funcs updates

ickshonpe committed:
check for `Node` not `NodeSize` removed components

ickshonpe committed:
fixed removal detection

ickshonpe committed:
remove rather than reuse old taffy nodes

ickshonpe committed:
also do a full update on window creation

ickshonpe committed:
`window_changed` check refactor

ickshonpe committed:
changed name of taffy key component to `UiKey`

ickshonpe committed:
Removed the insertion system

ickshonpe committed:
Cleaned up some comments, removed the `Clone` impl for `UiKey`.

ickshonpe committed:
Doc comments edit.

ickshonpe committed:
Merge branch 'main' into taffy-key-component

ickshonpe committed:
removed unnecessary changes

ickshonpe committed:
removed `is_null` method, cargo fmt --all

ickshonpe committed:
Renamed `UiKey` to `UiNodeId`, added type alias `NodeId` for `taffy::node::Node`

ickshonpe committed:
added `UiSurface` methods for insertion, and setting style and measure funcs

ickshonpe committed:
Merge branch 'main' into taffy-key-component

ickshonpe committed:
cargo fmt --all

ickshonpe committed:
Changed `UiNodeId` from a tuple struct to a stuct with a `taffy_id` field and `get` method.

Edited the doc comment for `insert_node`.

ickshonpe committed:
Merge branch 'store-taffy-keys-in-components' into taffy-key-component-merge

ickshonpe committed:
Renamed `UiNodeId` to `UiKey`, and moved its struct declaration to the `layout` module.

Renamed some variables to be consistant with the name change.

`UiSurface::insert_node` reuses the previous taffy node associated to the given `entity`, if one exists.

ickshonpe committed:
cargo fmt --all

ickshonpe committed:
Changed some paramter names to be more descriptive.

ickshonpe committed:
Removed `LayoutError` as it's no longer used.

ickshonpe commented:
Improved the implementation and reduced it down to just the minimum necessary changes without any extra systems. Should be ready to merge, but I'll make some new tests for `UiSurface`.

ickshonpe committed:
`layout` module:
Changed the ordering of parameters for the `from_style` and `LayoutContext::new` functions to put the most important parameter first.

`UiSurface`:
Added a test `uisurface_insert_and_remove_nodes`.

ickshonpe committed:
`Taffykey`:
Type alias for `taffy::node::Node`.

`UiSurface::get_key`
Method that can be used to lookup taffy nodes from `entity_to_taffy` but takes and returns a value rather than a borrow.

ickshonpe committed:
`Taffykey`:
Type alias for `taffy::node::Node`.

`UiSurface::get_key`
Method that can be used to lookup taffy nodes from `entity_to_taffy` but takes and returns a value rather than a borrow.

ickshonpe committed:
Merge branch 'store-taffy-keys-in-components' of https://github.com/ickshonpe/bevy into store-taffy-keys-in-components

ickshonpe committed:
cargo fmt

ickshonpe committed:
moved window handling in `ui_layout_system` outside of the update logic.

alice-i-cecile committed:
Merge branch 'main' into store-taffy-keys-in-components

mockersf commented:
It's increasing performances at the cost of memory. Not sure it's worth it in this case...

cart commented:
I do like the idea of breaking this out from a centralized hashmap (and perf wins are good), but having multiple sources of truth that need syncing gives me some pause. I'll need to give this some thought, which sadly will need to wait until 0.12.

ickshonpe committed:
Merge branch 'main' into store-taffy-keys-in-components

ickshonpe committed:
Cleaned up left over comments from merge

ickshonpe commented:
Better to implement the taffy tree traits.
